projects/driver_controls/src/can_fsm.c

```diff
@@ -3,36 +3,52 @@
 #include "input_event.h"
 #include "log.h"
 
+static FSM *s_can_fsm;
+
 FSM_DECLARE_STATE(state_can_transmit);
 
 FSM_STATE_TRANSITION(state_can_transmit) {
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_POWER, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_PEDAL, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_DIRECTION_SELECTOR, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_TURN_SIGNAL, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_HAZARD_LIGHT, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_MECHANICAL_BRAKE, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_HORN, state_can_transmit);
-  FSM_ADD_TRANSITION(INPUT_EVENT_CAN_ID_PUSH_TO_TALK, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_POWER, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_PEDAL, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_DIRECTION_SELECTOR, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_TURN_SIGNAL, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_HAZARD_LIGHT, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_MECHANICAL_BRAKE, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_HORN, state_can_transmit);
+  FSM_ADD_TRANSITION(CAN_DEVICE_ID_PUSH_TO_TALK, state_can_transmit);
 }
 
 static void prv_transmit_data(FSM *fsm, const Event *e, void *context) {
-  InputEventData *data = (InputEventData *)&e->data;
-
-  printf("Device = %d, State = %d, Data = %d\n", e->id, data->components.state,
-         data->components.data);
+  printf("Device = %d, State = %d, Data = %d\n",
+          e->id - NUM_INPUT_EVENT,
+          e->data >> 12,
+          e->data & 0xFFF);
 }
 
 StatusCode can_fsm_init(FSM *fsm) {
+  s_can_fsm = fsm;
+
   fsm_state_init(state_can_transmit, prv_transmit_data);
 
-  void *context = event_arbiter_add_fsm(fsm, NULL);
+  fsm_init(fsm, "can_fsm", &state_can_transmit, NULL);
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode can_fsm_transmit(CANDeviceID device_id, uint8_t device_state, uint16_t device_data) {
+  union EventData {
+    uint16_t raw;
+    struct {
+      uint16_t data : 12;
+      uint8_t state : 4;
+    } components;
+  } data;
 
-  if (context == NULL) {
-    return status_code(STATUS_CODE_RESOURCE_EXHAUSTED);
-  }
+  data.components.data = device_data;
+  data.components.state = device_state;
 
-  fsm_init(fsm, "can_fsm", &state_can_transmit, context);
+  Event e = {.id = device_id, .data = data.raw };
+  fsm_process_event(s_can_fsm, &e);
```

Comment:
This is equivalent to just calling a function. :|

At this point, I feel like using the FSM is pointless for you and doesn't really take advantage of the 64-bit data size.

Reply:
Sorry, I thought I got rid of that in the merge